matrix_functions.py

```diff
@@ -381,24 +415,63 @@ def _matrix_inverse_root_eigen(
     # compute eigendecomposition and compute minimum eigenvalue
     L, Q = _eigh_eigenvalue_decomposition(
         A_ridge,
-        retry_double_precision=retry_double_precision,
-        eigendecomposition_offload_device=eigendecomposition_offload_device,
+        retry_double_precision=root_inv_config.retry_double_precision,
+        eigendecomposition_offload_device=root_inv_config.eigendecomposition_offload_device,
     )
-    lambda_min = torch.min(L)
 
-    # make eigenvalues > 0 (if necessary)
+    power_L = scale_and_pow_eigenvalues(
+        L,
+        root,
+        epsilon=epsilon,
+        noninvertible_handling_config=root_inv_config.noninvertible_handling_config,
+    )
+
+    # compute inverse preconditioner
+    X = Q * power_L.unsqueeze(0) @ Q.T
+
+    return X, L, Q
 
-    if enhance_stability:
-        L += -torch.minimum(lambda_min - epsilon, torch.as_tensor(0.0))
+
+def scale_and_pow_eigenvalues(
+    L: Tensor,
+    root: Fraction,
+    epsilon: float = 0.0,
+    noninvertible_handling_config: NonInvertibleHandlingConfig = DefaultRegularizationConfig,
+) -> Tensor:
+    if isinstance(noninvertible_handling_config, PseudoInverseConfig):
+        assert epsilon == 0.0, f"{epsilon=} must be 0.0 when using pseudo-inverse!"
+
+        # Filter the eigenvalues based on the numerical rank of the matrix
+        # The procedure below mimics the steps described in the documentation of
+        # https://pytorch.org/docs/stable/generated/torch.linalg.matrix_rank.html
+        if noninvertible_handling_config.rank_rtol is None:
+            rtol = L.numel() * torch.finfo(L.dtype).eps
+        else:
+            rtol = noninvertible_handling_config.rank_rtol
+        spectrum_cutoff = max(
+            noninvertible_handling_config.rank_atol, rtol * L.max().relu().item()
+        )
+        power_L = torch.where(
+            L <= spectrum_cutoff,
+            torch.zeros_like(L),
+            L.pow(torch.as_tensor(-1.0 / root)),
+        )
     else:
-        L += -torch.minimum(lambda_min, torch.as_tensor(0.0))
-        # and add the epsilon
-        L += epsilon
+        assert isinstance(noninvertible_handling_config, RegularizationConfig)
 
-    # compute the matrix inverse root
-    X = Q * L.pow(torch.as_tensor(-1.0 / root)).unsqueeze(0) @ Q.T
+        lambda_min = torch.min(L)
 
-    return X, L, Q
+        # make eigenvalues > 0 (if necessary)
+        if noninvertible_handling_config.add_epsilon_before_computation:
+            L += -torch.minimum(lambda_min - epsilon, torch.as_tensor(0.0))
```

Comment:
Any reason to not write
```suggestion
            L -= torch.minimum(lambda_min - epsilon, torch.as_tensor(0.0))
```
?

Reply:
No, I just copied the code as it was before.